presto-common/src/main/java/com/facebook/presto/common/Page.java

```diff
@@ -318,6 +318,28 @@ public Page getPositions(int[] retainedPositions, int offset, int length)
         return wrapBlocksWithoutCopy(length, blocks);
     }
 
+    public Page copyPositions(int[] retainedPositions, int offset, int length)
+    {
+        requireNonNull(retainedPositions, "retainedPositions is null");
+
+        Block[] blocks = new Block[this.blocks.length];
+        for (int i = 0; i < blocks.length; i++) {
+            blocks[i] = this.blocks[i].copyPositions(retainedPositions, offset, length);
+        }
+        return wrapBlocksWithoutCopy(length, blocks);
+    }
+
+    public Page extractChannels(int[] channels)
+    {
+        requireNonNull(channels, "channels is null");
+
```

Comment:
Should we validate the channel ranges?

Reply:
I thought about that, but decided against it because none of the usage sites that were previously doing this kind of operation before we’re checking the ranges (with one exception that checked that all indexes were positive in its constructor, and still does).

My worry would be about adding more overhead in checking per call rather than once during construction that the JIT may or may not be able to leverage to elide range checks during access anyway, so I left it out.